### PR TITLE
🌱 Add KCP conditions, split reconcileHealth into preflight and reconcileEtcdMembers, make both use conditions

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
Forward port of https://github.com/kubernetes-sigs/cluster-api/pull/3900

This PR adds KCP specific conditions to KCP itself and to machines providing visibility in:
- the status of control plane components/static pods generated by kubeadm
- the status of etcd members

The initial PR attempt was based on the idea of assigning to `reconcileHealth` the additional responsibility of computing conditions, however, after discovering a few problems during tests it was decided to propose a bigger refactor that
- assigns the responsibility of upgrading conditions to a new `reconcileControlPlaneConditions` func and to `workloadcluster.UpdateStaticPodConditions` and `workloadcluster.UpdateEtcdConditions`
- replaces `reconcileHealth` with `preflightChecks`, now based on conditions state.
- introduces `reconcileEtcdMembers`, for taking care of etcd members without machines (before this was improperly implemented in `reconcileHealth`

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3909

### CODE REVIEW

**Comment by fabriziopandini** [comment]
/hold 
for https://github.com/kubernetes-sigs/cluster-api/pull/3884 to merge

**Comment by vincepri** [comment]
@fabriziopandini Can we unhold this now?

**Comment by fabriziopandini** [comment]
Sure! I forgot to unhold after rebasing earlier today
/unhold

File: controlplane/kubeadm/controllers/controller.go [hunk 1]
```diff
@@ -23,22 +23,15 @@ import (
 
 	"github.com/blang/semver"
 	"github.com/pkg/errors"
+
```
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/controller.go:26]:
```suggestion
```

File: controlplane/kubeadm/internal/etcd/util/util.go [hunk 2]
```diff
@@ -30,11 +31,16 @@ func MemberForName(members []*etcd.Member, name string) *etcd.Member {
...
+		names = append(names, m.Name)
 	}
-	return set
+	return names
+}
+
+func MemberEqual(members1, members2 []*etcd.Member) bool {
+	names1 := sets.NewString(MemberNames(members1)...)
+	names2 := sets.NewString(MemberNames(members2)...)
+	return names1.Equal(names2)
```
Comment (sedefsavas) [inline controlplane/kubeadm/internal/etcd/util/util.go:45]: Are these come always in the same order? NewString does not guarantee order.

**Review by sedefsavas — COMMENTED** [review]
/lgtm

File: controlplane/kubeadm/internal/etcd/util/util.go (hunk 2 above)
Comment (vincepri) [reply controlplane/kubeadm/internal/etcd/util/util.go:45]: @fabriziopandini

Comment (fabriziopandini) [reply controlplane/kubeadm/internal/etcd/util/util.go:45]: NewString returns a set.String, ordering is not relevant for this type

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**